refactor / kernel: Move non-gArgs chainparams functionality to kernel

sedited opened the pull request:
This pull request is part of the `libbitcoinkernel` project https://github.com/bitcoin/bitcoin/issues/24303 https://github.com/bitcoin/bitcoin/projects/18 and more specifically its "Step 2: Decouple most non-consensus code from libbitcoinkernel". dongcarl is the original author of this patchset, these commits were taken from https://github.com/dongcarl/bitcoin/tree/2022-03-libbitcoinkernel-chainparams-args-only. 

#### Context 

The bitcoin kernel library currently relies on code containing user configurations through the `ArgsManager`. This is not optimal, since as a stand-alone library it should not rely on bitcoind's argument parsing logic. Instead, its interfaces should accept control and options structs that control the kernel library's desired configuration.


Similar work towards decoupling the `ArgsManager` from the kernel has been done in 
https://github.com/bitcoin/bitcoin/pull/25290, https://github.com/bitcoin/bitcoin/pull/25487, https://github.com/bitcoin/bitcoin/pull/25527 and https://github.com/bitcoin/bitcoin/pull/25862.

#### Changes

By moving the `CChainParams` class definition into the kernel and giving it new factory functions `CChainParams::{RegTest,SigNet,Main,TestNet}`it can be constructed without an `ArgsManager` reference, unlike the current factory function `CreateChainParams`.

The first few commits remove uses of `ArgsManager` within `CChainParams`. Then the `CChainParams` definition is moved to a new file in the `kernel/` subdirectory.

dongcarl committed:
Decouple SigNetChainParams from ArgsManager

SigNet chain params can now be initialized by configuring a
SigNetOptions struct, or with ArgsManager. This offers an interface for
creating SigNetChainParams without a gArgs object.

dongcarl committed:
Decouple RegTestChainParams from ArgsManager

RegTest chain params can now be initialized by configuring a
RegTestOptions struct, or with ArgsManager. This offers an interface for
creating RegTestChainParams without a gArgs object.

dongcarl committed:
Remove UpdateVersionBitsParameters

Moves setting struct member fields from a function to its call site.
This improves readability by surfacing the code.

dongcarl committed:
Add factory functions for Main/Test/Sig/Reg chainparams

This normalizes the behavior of initializing Main/Test/Sig/Reg
chainparams with RegTest/SigNet chainparams. These factory functions can
also easily be used from a context without an instantiated ArgsManager,
e.g. from libbitcoin kernel code, unlike the existing CreateChainParams
method.

dongcarl committed:
Split non/kernel chainparams

Moves chainparams code not using the ArgsManager to the kernel.

Subsequently use the kernel chainparams header now where possible in
order to further decouple chainparams call sites from gArgs.

dongcarl commented on src/chainparams.h:
```diff
@@ -146,6 +146,32 @@ class CChainParams
+};
+
+/**
+ * Return SigNetOptions filled with default values.
```
I think it's not really "SigNetOptions filled with default values" but rather "Options for the default signet". I remember @theuni made a good point offline that it almost never makes sense to override _some_ of the options for the default signet.

dongcarl commented on src/chainparams.cpp:
```diff
@@ -415,7 +415,8 @@ std::unique_ptr<const CChainParams> CreateSignetChainParams(const ArgsManager& a
 class CRegTestParams : public CChainParams
```
Ping @ajtowns, could you take a peek? 9929969d3e95197499f2c81c01702f11aa52ff0d

dongcarl commented on src/kernel/chainparamsbase.cpp:
```diff
@@ -0,0 +1,43 @@
+    throw std::runtime_error(strprintf("%s: Unknown chain %s.", __func__, chain));
+}
+
+static std::unique_ptr<CBaseChainParams> globalChainBaseParams;
```
Hmmm... Been a while since I last took a look but how hard would it be to have this as a member of `kernel::Context`? What does the caller graph look like?

It might be the case that this is a situation where it'd be too intrusive to have the N I C E C O D E as discussed here: https://github.com/bitcoin/bitcoin/pull/25065#pullrequestreview-985450940

TheCharlatan replied:
The caller graph is kind of annoying, it boils down to users of `SelectBaseParams` (like bitcoin-cli), and more importantly `SelectParams`, which is used all over the place (I count 23 separate files as call site). If I understand the current usage of `kernel::Context` correctly, this would mean that we'd have to construct the context for nearly each call site separately, and many in places where it was previously not required. I currently tend towards handling this in a separate refactor attempt to get rid of these param globals in general.

ajtowns replied:
You could maybe split it up? Have chainparams provide `SelectParams(ArgsManager&)` and use that instead of `SelectParams(args.GetChainName())` which covers all the cases outside of test/ and test/fuzz/. Then most of the test or test/fuzz cases already hardcode the network, and could be replaced by `kernel::SelectParams(RegTestBase(), RegTest(regtestopts))` that just initializes the globals?

TheCharlatan replied:
I've had no luck coming up with a nice pattern to get this split done without blowing everything up :/
I might try it again at a later point in time by first create an ugly solution and then see if I can mold it into something nice.

TheCharlatan replied:
I looked into this again. The underlying problem seems to be that the kernel library still depends on the `ArgsManager`, which in turn depends on the global. I'd rather not make this change devolve into changing `ArgsManager` internals and behaviour, so I'll defer refactoring this global to once the `ArgsManager` is removed from the kernel library.

dongcarl commented on src/bitcoin-cli.cpp:
```diff
@@ -67,10 +67,10 @@ static void SetupCliArgs(ArgsManager& argsman)
-    const auto testnetBaseParams = CreateBaseChainParams(CBaseChainParams::TESTNET);
-    const auto signetBaseParams = CreateBaseChainParams(CBaseChainParams::SIGNET);
-    const auto regtestBaseParams = CreateBaseChainParams(CBaseChainParams::REGTEST);
+    const auto defaultBaseParams = kernel::CreateBaseChainParams(CBaseChainParams::MAIN);
```
nit: Could use `using` declarations in `.cpp` files (if you want)

TheCharlatan replied:
Seems to me like `using` declarations are not used for internal namespaces in the code so far, so I chose not to.

dongcarl reviewed (commented):
Concept ACK

Did a first-pass review. Perhaps @ajtowns and @kallewoof can help w/re the activation/chainparams/signet bits!

ajtowns commented on src/chainparams.cpp:
```diff
@@ -272,23 +272,32 @@ class CTestNetParams : public CChainParams {
+        std::vector<uint8_t> bin{options.challenge};
+        vSeeds = options.seeds;
 
+        if (bin == SIGNET_DEFAULT_CHALLENGE) {
```
I think if you wanted exactly the same behaviour as current, it would be better to write:

```c++
    if (options.challenge.empty()) { /* use defaults */ } else { /* use values from options */ }
```

TheCharlatan replied:
Done in 7d7900edeff6af18c3bb2c11475ac26414ed78d2

ajtowns reviewed (commented):
Concept ACK. Having simple options structs as a buffer between code and `gArgs` is great IMO.

Some comments below; code implementing them, and some other ideas at https://github.com/ajtowns/bitcoin/commits/202209-kernelchainparams

ryanofsky reviewed (commented):
Just reviewed the first commit so far, but already I think some things are a little overcomplicated and potentially buggy. I would suggest having a 1:1 correspondence between arguments and options, treating empty arguments as unset options, and keeping PR a pure refactoring which does not change behavior.

Started review (will update list below with progress).
- [X] 8c0a94f1d56345c45304319e72b7f78626e47061 Decouple SigNetChainParams from ArgsManager (1/7)
- [ ] 9929969d3e95197499f2c81c01702f11aa52ff0d Decouple RegTestChainParams from ArgsManager (2/7)
- [ ] 4ca849d6bc2d617babdcb870ca1de4376917c11d Remove UpdateVersionBitsParameters (3/7)
- [ ] 5d70983a84c9e09d315bd19296a08c26570b3b9a Add factory functions for Main/Test chainparams (4/7)
- [ ] c00f4f7599b89df8196287176e743a99da1de67f split non/kernel chainparamsbase (5/7)
- [ ] e461bd23908802ad806cdb25a4cfe9e23149dc57 Split non/kernel chainparams (6/7)
- [ ] 9714fe9160282dc36269b1d16d57dfdd48d0a1f1 Wrap kernel chainparamsbase functions in kernel namespace (7/7)

ryanofsky commented on src/chainparams.cpp:
```diff
@@ -382,11 +381,39 @@ class SigNetParams : public CChainParams {
+        return CreateSignetChainParams(SigNetOptions{ParseHex(signet_challenge[0]), seeds});
+    } else {
+        auto opts = GetDefaultSignetOptions();
+        if (!seeds.empty()) {
```
> I think that is taking it a bit far. 

Right I'm just talking about a minor change to the `SigNetOptions` struct that would simplify the changes here.

The only thing this PR really needs to do is get rid of `ArgsManager` usage in this part of the code, replacing `ArgsManager` lookups with `SigNetOptions` member accesses. I suggested a slightly different version of the `SigNetOptions` struct above that would let it provide the same information `ArgsManager` provides, in a 1:1 way so less code has to change, and so the struct is easier for callers to fill and customize.

> But if that were really the case, why not just pass a fully fleshed out const CChainParams& to the kernel in the first place?

This is fine, and I think it basically already happens later after `CChainParams` are initialized. I'm just looking at the earlier code used to initialize `CChainParams`. I do think alternate approaches are possible. Like instead of `SigNetOptions` being used to initialize `CChainParams` and have  `CChainParams` contain dynamic settings determined at runtime, `CChainParams` could just contain static hardcoded settings, and code that uses `CChainParams` could use combine it with runtime settings. This already happens with chain parameters like `nMinimumChainWork` that can be overridden on the command line.

> In that case, kernel/chainparams.h could be a headers only module, with the mainnet/signet/etc setup code staying in chainparams.cpp. I'm not sure if I'm missing a design goal here though.

I'm not sure exactly what this would look like but it does sound fine. Maybe similar to what I was describing above making `CChainParams` not vary depending on runtime options.

These alternate approaches are just bigger than changes I was suggesting above. All I was suggesting above was replacing ArgsManager uses with SigNetOptions in a 1:1 way instead of doing more complicated mappings and creating another set of defaults and making it harder to initialize.

ajtowns replied:
> I think that is taking it a bit far.

Yeah -- I think I agree, I just don't know why I agree? I was more hoping to understand why that would be taking it a bit far?

ryanofsky replied:
Rereading I think we might be talking past each other. You mentioned things like "translating to text" which I'm not suggesting. I'm literally just agreeing with your initial suggestion `if (options.challenge.empty()) { /* use defaults */ } else { /* use values from options */ }` and saying how you can tweak the options struct definition so the same approach works for seeds as well, and the options struct does not need to contain any new or different information than what ArgsManager provides.

Later you mentioned passing "a fully fleshed out const CChainParams&" and having a "headers only module" and I think these changes sound ok, but maybe involve getting rid of the options struct or not using it here, and I'm just proposing tweaking the struct a little.

ajtowns replied:
@ryanofsky Yeah, I understand what you were saying, and tweaking the approach in this PR feels sensible to me too; I just don't feel like I intellectually understand the tradeoffs here, so was hoping thinking about more extreme alternatives might be revealing.

Three approaches:
 * `kernel::RegTestOptions` just contains the exact results from gArgs (so a vector of strings, not a map of enums etc)
 * what's implemented in this PR, with some tweaks
 * keep all the mainnet/testnet/etc defaults out of kernel, and just give it a prefilled CChainParams

I think this PR is better than the first option, because text is annoying, and the tweaks make it even better. I don't have a handle on the tradeoffs for the third approach, though.

ryanofsky replied:
I think there is still a misunderstanding. I'm _not_ suggesting using text instead of enums or using hex strings instead of binary strings in the Options struct.

So I don't think I'm suggesting any of the approaches you listed above. I'm suggesting a fourth approach: `kernel::RegTestOptions` contains a 1:1 representation of information in `gArgs`. If a gArgs option is unset, corresponding Options parameter is unset or empty. If gArgs option is valid, corresponding options member contains the parsed option value. No special calls are needed to fill the options struct with defaults, because it just does the default thing by default and can be easily customized if desired. `bitcoind` behavior is unchanged and less code has to change because the options struct can be used as a straightforward replacement for ArgsManager calls.

I think if you want to go further and remove more code from the kernel, it could be good or bad, and the tradeoffs you are thinking about come into play. But I think the only thing we actually want to remove right now is ArgsManager. I don't think we need to something like removing preset chain parameters.

ajtowns replied:
Right -- I'm describing your (and my) suggestions as "some tweaks" to this PR. (They seem like obviously good ideas to me, so I was glossing over them in favour of the things I'm confused about)

ryanofsky replied:
Makes sense!

TheCharlatan commented on src/kernel/chainparams.h:
```diff
@@ -0,0 +1,184 @@
+struct RegTestOptions {
+    std::unordered_map<Consensus::DeploymentPos, VersionBitsParameters> version_bits_parameters{};
+    std::unordered_map<Activations, int> activation_heights{};
+    uint64_t prune_after_height{1000};
```
This is corrected in 795b6fb83d8eee9a7f47629b274feca066fc17fb

TheCharlatan commented:
Thank you all for the reviews, I reworked my commits to include your suggestions. 

@ajtowns I molded most of your suggested code into the existing commits. I did not include the complete refactor of the `Consensus::BuriedDeployment` though.

@ryanofsky the options structs are now instantiated with default values and read from the `ArgsManager` in dedicated functions. Since I decided to move the options structs into `CChainParams`, I did not move these dedicated functions into a separate file.

ajtowns commented on src/kernel/chainparams.cpp:
```diff
@@ -0,0 +1,510 @@
+                .nTxCount = 1903567,
+                .dTxRate  = 0.02336701143027275,
+            };
+            consensus.signet_challenge = ParseHex("512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430210359ef5021964fe22d6f8e05b2463c9540ce96883fe3b278760f048f5189f2e6c452ae");
```
```suggestion
            consensus.signet_challenge = ParseHex("512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430210359ef5021964fe22d6f8e05b2463c9540ce96883fe3b278760f048f5189f2e6c452ae");
            if (vSeeds.empty()) {
                vSeeds.push_back("seed.signet.bitcoin.sprovoost.nl.");
                vSeeds.push_back("178.128.221.177");
                vSeeds.push_back("v7ajjeirttkbnt32wpy3c6w3emwnfr3fkla7hpxcfokr3ysd3kqtzmqd.onion:38333");
            }
```

along with defaulting `options.seeds` to empty. Otherwise just overriding `options.challenge` will have you connecting to the default seeds which would be annoying.

TheCharlatan replied:
Thanks for review, I pushed another potential solution using constants. Let me know if you like something more akin to your solution, or fully using defaults better.

ajtowns replied:
Currently, if you run`bitcoind -signet -signetchallenge=51` then you start up with no seed nodes; with your current patch, you'll instead connect to the default signet seednodes, and then reject them for having the wrong network magic. That just seems buggy to me? Putting these defaults into the code rather than the structure definition seems like the easiest way of making that work sensibly to me.

TheCharlatan replied:
I thought I changed the behaviour now: It does not populate the seeds if the passed in option seeds are default and the signet challenge is non-default. This is still kind of confusing in the code though, so please let me know if you find that inferior.

ajtowns replied:
Err, somehow I missed that entirely? Whoops.

It's not the way I'd do it, but as long as it works, I don't really mind.

Another approach would be to have `SigNetOptions` default to an empty challenge and empty seed vector, but define a `const SigNetOptions GlobalDefaultSigNet { DEFAULT_CHALLENGE, DEFAULT_SEEDS };` with the global defaults, then make `ReadSigNetArgs` do:

```c++
if (args.IsArgSet("-signetchallenge")) {
    ...
    options.challenge = ParseHex(signet_challenge[0]);
} else {
    options = GlobalDefaultSigNet;
}
if (args.IsArgSet("-signetseednode")) {
    options.seeds = args.GetArgs("-signetseednode");
}
```

then you don't have to special case challenge/seeds in `SigNetParams(options)`, though you'd still be special casing min chain work, assume valid, etc.

TheCharlatan replied:
Commit "Decouple SigNetChainParams from ArgsManager" e893056ed68d02ef94f7e94fc6287217610723a4 now uses a struct with optionals as proposed here.

I tried implementing similar optionals for the RegTestOptions in commit "Douple RegTestChainParams from ArgsManager" 6d57a29df85277964edb21ee251b395e4eae1ef6, but it felt un-ergonomic.

ajtowns replied:
Using optionals for signet but not regtest makes sense to me.

jamesob commented:
Ping @TheCharlatan for a rebase here.

TheCharlatan commented:
> Ping @TheCharlatan for a rebase here.

Will get to it Monday.

TheCharlatan commented:
Rebased.

fanquake commented:
Looks like all three test failures are the same "Blockchain is too short for pruning":

https://cirrus-ci.com/task/5072222193188864?logs=functional_tests#L2444
```bash
 test  2023-01-17T22:29:12.633000Z TestFramework (ERROR): JSONRPC error 
                                   Traceback (most recent call last):
                                     File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\test_framework\test_framework.py", line 134, in main
                                       self.run_test()
                                     File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\feature_index_prune.py", line 73, in run_test
                                       pruneheight_new = node.pruneblockchain(400)
                                     File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\test_framework\coverage.py", line 49, in __call__
                                       return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                                     File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\test_framework\authproxy.py", line 146, in __call__
                                       raise JSONRPCException(response['error'], status)
                                   test_framework.authproxy.JSONRPCException: Blockchain is too short for pruning. (-1)
```

https://cirrus-ci.com/task/5037121606516736?logs=ci#L3282
```bash
 test  2023-01-17T22:14:21.962000Z TestFramework (ERROR): JSONRPC error 
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 134, in main
                                       self.run_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_utxocache.py", line 150, in run_test
                                       self.test_flush()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/interface_usdt_utxocache.py", line 397, in test_flush
                                       self.nodes[0].pruneblockchain(315)
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/coverage.py", line 49, in __call__
                                       return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/authproxy.py", line 146, in __call__
                                       raise JSONRPCException(response['error'], status)
                                   test_framework.authproxy.JSONRPCException: Blockchain is too short for pruning. (-1)
```

TheCharlatan commented:
Thanks for the heads up fanquake, there was a bug that is now fixed in 1ad1c68. I'm not sure about the one remaining test failure.

ajtowns commented on src/chainparams.cpp:
```diff
@@ -382,11 +378,25 @@ class SigNetParams : public CChainParams {
     }
 };
 
+void ReadSigNetArgs(const ArgsManager& args, CChainParams::SigNetOptions& options) {
```
Seems more ergonomic to write:

```c++
CChainParams::SigNetOptions ReadSigNetArgs(const ArgsManager& args)
{
    ...
```

but maybe this is just deliberately keeping with the same pattern as per https://github.com/bitcoin/bitcoin/pull/26883/files#r1068247937

TheCharlatan replied:
Yes, also keeping it the same as in: https://github.com/bitcoin/bitcoin/pull/25862/files#diff-4e268aeb074a176689aace31957a889c1f39c909e6678ca7be8707bda1a7be56R11

ajtowns commented on src/chainparams.cpp:
```diff
@@ -436,11 +437,33 @@ class CRegTestParams : public CChainParams
         m_assumed_chain_state_size = 0;
 
-        UpdateActivationParametersFromArgs(args);
+        for (const auto& [name, height] : opts.activation_heights) {
```
`name` is really `deployment_id` (`dep` for short)?

TheCharlatan replied:
I see, I'll use `dep`, because that's what is used elsewhere in the code.

ajtowns reviewed (commented):
Approach ACK f54210f50ec0cd44f026988308d975397a948d8c

TheCharlatan commented:
Thanks for the review. [diff](https://github.com/bitcoin/bitcoin/compare/f54210f50ec0cd44f026988308d975397a948d8c..1f8c242e07a0842a3c7cc8735a6dde3e92327bb4)

TheCharlatan commented:
Rebased.

ryanofsky commented on src/bitcoin-chainstate.cpp:
```diff
@@ -51,8 +51,9 @@ int main(int argc, char* argv[])
     // SETUP: Misc Globals
-    SelectParams(CBaseChainParams::MAIN);
-    const CChainParams& chainparams = Params();
+    SetGlobalBaseParams(CBaseChainParams::MAIN);
```
In commit "Split non/kernel chainparams" (4820450688e371a99d66cfad24714fa6e041e43e)

Is calling SetGlobalBaseParams() actually necessary here? I'm guessing something breaks if isn't added but it's not clear what. It'd be good drop the call or change the generic comment above "// SETUP: Misc Globals" into a specific comment about why it's needed currently, and maybe how it can go away in the future.

TheCharlatan replied:
This is no longer relevant.

ryanofsky reviewed (commented):
Partial code review ACK for 2ec5cdb9e4884b75623a0cc4328ac78bd5d075c9, which is the fourth commit. The first 4 commits resolve all the previous issues that were discussed and look ready to merge. I did leave some new comments about code style, but they aren't important and can be ignored.

The last 3 commits seem more half-baked to me, and might be better to split off. They require a different type of review anyway since they are moves and renames, not actual code changes. I left a more detailed comment below but basically I think the chainparams commits make sense and the chainparamsbase commits don't. So an alternative to splitting up this PR could be to drop the 5th and 7th commits.

- [X] 2795ff9cc40383f74223736c3b96515faea37410 Decouple SigNetChainParams from ArgsManager (1/7)
- [X] 1cfccc38dd5d39a9e23a19b94d6ba1a7bc4c48dc Decouple RegTestChainParams from ArgsManager (2/7)
- [X] 96e01ecb18f926355fdad4da3d30d48788641082 Remove UpdateVersionBitsParameters (3/7)
- [X] 2ec5cdb9e4884b75623a0cc4328ac78bd5d075c9 Add factory functions for Main/Test/Sig/Reg chainparams (4/7)
- [ ] 2f2216dee7a683638e6a3e966142f50f5de7df78 split non/kernel chainparamsbase (5/7)
- [X] 4820450688e371a99d66cfad24714fa6e041e43e Split non/kernel chainparams (6/7)
- [ ] a1db3966d1c4f4213d7bfd6be0ea7fb308fe4e11 Wrap kernel chainparamsbase functions in kernel namespace (7/7)

TheCharlatan commented on src/chainparams.cpp:
```diff
@@ -436,11 +437,33 @@ class CRegTestParams : public CChainParams
         m_assumed_chain_state_size = 0;
 
-        UpdateActivationParametersFromArgs(args);
+        for (const auto& [dep, height] : opts.activation_heights) {
```
I'll keep this as is, to me this reads easier. We also do use the options struct before to read the fastprune flag.

TheCharlatan commented:
Thank you for the review! These are valuable and I hope they allow me to improve future commits. I'm pushing changes addressing comments up to the 4th commit now. 

Updated a1db3966d1c4f4213d7bfd6be0ea7fb308fe4e11 -> 65d026b47270c2fb9d2dab7f3d448b4abc37b93a ([tc/2022-09-libbitcoinkernel-chainparams-args_0](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_0) -> [tc/2022-09-libbitcoinkernel-chainparams-args_1](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_0..tc/2022-09-libbitcoinkernel-chainparams-args_1)) to address some of @ryanofsky 's comments.

> The last 3 commits seem more half-baked to me, and might be better to split off.

I'll now move to assessing how to proceed here now. I agree though that chainparamsbase could go into another PR altogether. The current approach is kind of ugly.

TheCharlatan commented on src/chainparams.cpp:
```diff
@@ -582,20 +582,39 @@ const CChainParams &Params() {
     return *globalChainParams;
 }
 
+std::unique_ptr<const CChainParams> CChainParams::SigNet(const CChainParams::SigNetOptions& options) {
```
Thank you for spelling this out, I updated the description.

TheCharlatan commented on src/chainparams.cpp:
```diff
@@ -497,41 +520,32 @@ class CRegTestParams : public CChainParams
+    options.fastprune = args.GetBoolArg("-fastprune", false);
+
+    for (const std::string& arg : args.GetArgs("-testactivationheight"))
+    {
```
Good catch, updated. [diff](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_0..tc/2022-09-libbitcoinkernel-chainparams-args_1)

TheCharlatan commented on src/deploymentinfo.cpp:
```diff
@@ -34,3 +34,18 @@ std::string DeploymentName(Consensus::BuriedDeployment dep)
     return "";
 }
+
+std::optional<Consensus::BuriedDeployment> GetBuriedDeployment(const std::string& deployment_name) {
```
Good point, updated. [diff](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_0..tc/2022-09-libbitcoinkernel-chainparams-args_1)

TheCharlatan commented:
> The last 3 commits seem more half-baked to me, and might be better to split off.

Ok, I kept the 5th commit, "split non/kernel chainparamsbase", but limited the best to just moving the declaration of the `CBaseChainParams` class to the kernel. I also dropped the 7th commit "wrap kernel chainparamsbase functions in kernel namespace".


Updated 65d026b47270c2fb9d2dab7f3d448b4abc37b93a -> 067b727a9741ddabf44f6efd5125a9c8fbea0702 ([tc/2022-09-libbitcoinkernel-chainparams-args_1](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_1) -> [tc/2022-09-libbitcoinkernel-chainparams-args_2](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_1..tc/2022-09-libbitcoinkernel-chainparams-args_2)) to reduce the refactor of chainparamsbase as suggested by @ryanofsky.

TheCharlatan commented on src/bitcoin-chainstate.cpp:
```diff
@@ -51,8 +51,9 @@ int main(int argc, char* argv[])
-    const CChainParams& chainparams = Params();
+    SetGlobalBaseParams(CBaseChainParams::MAIN);
+    auto chainparams_ptr = CChainParams::Main();
+    const CChainParams& chainparams = *chainparams_ptr;
```
Thanks, updated. [Diff](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_1..tc/2022-09-libbitcoinkernel-chainparams-args_2)

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/26177#discussion_r1126829031

> Thanks, updated. [Diff](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_1..tc/2022-09-libbitcoinkernel-chainparams-args_2)

This looks the same as before. I was suggesting dropping the `chainparams_ptr` variable and just writing `*chainparams` directly below.

TheCharlatan replied:
I dropped this by mistake. Included now.

TheCharlatan commented:
Updated 067b727a9741ddabf44f6efd5125a9c8fbea0702 -> f9e3c91f4d482b0c74ef899056d5a2c63e67e1fb ([tc/2022-09-libbitcoinkernel-chainparams-args_2](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_2) -> [tc/2022-09-libbitcoinkernel-chainparams-args_3](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_3), [compare](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_2..tc/2022-09-libbitcoinkernel-chainparams-args_3)) to undo an unnecessary order change in the makefile and to re-add a newline at the end of a file.

sedited committed:
refactor: Don't use global chainparams in chainstatemanager method

The chainstatemanager m_options.chainparams member variable gets its
value from the global chainparams in init.cpp. This allows
validation.cpp to only include the the kernel chainparams file.

TheCharlatan commented:
Updated f9e3c91f4d482b0c74ef899056d5a2c63e67e1fb -> c44983faa02b4845cd7605abe99c332980a07af1 ([tc/2022-09-libbitcoinkernel-chainparams-args_3](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_3) -> [tc/2022-09-libbitcoinkernel-chainparams-args_4](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_4), [compare](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_3..tc/2022-09-libbitcoinkernel-chainparams-args_4)) to rebase on latest master for fixing an include issue causing CI build failures, add a commit to fix the regression introducing said include issue, and update the copyright header on `kernel/chainparamsbase.h`.

ryanofsky commented on src/kernel/chainparamsbase.h:
```diff
@@ -0,0 +1,39 @@
+ * CBaseChainParams defines the base parameters (shared between bitcoin-cli and bitcoind)
+ * of a given instance of the Bitcoin system.
+ */
+class CBaseChainParams
```
In commit "split non/kernel chainparamsbase" (341be7fa680e9dcdbaa46db0194833b5d0229394)

I still don't think it makes sense to have `CBaseChainParams` in the kernel. Like the comment above says, the point of the `CBaseChainParams` class and the reason it is separate from `CChainParams` is that it is shared between bitcoin-cli and bitcoind. I think of `CBaseChainParams` as client parameters, and `CChainParams` as server parameters. Practically speaking, `bitcoin-cli` or other clients that needs basic information like network datadir locations or RPC port numbers should not be required to link against the kernel.

The justification for this change given in the commit message about allowing "kernel chainparams to be instantiated" also does not make sense to me, because chainparamsbase is part of libbitcoin_util, and it is fine for kernel code to depend on libbitcoin_util if we are going by https://github.com/bitcoin/bitcoin/blob/master/doc/design/libraries.md#dependencies

Previous comment about this was: https://github.com/bitcoin/bitcoin/pull/26177#discussion_r1126765676

TheCharlatan replied:
This makes sense, I wasn't connecting the fact that the util library includes `chainparamsbase` anyway with segregating the kernel code. I dropped the commit.

ryanofsky reviewed (commented):
Partial code review ACK c44983faa02b4845cd7605abe99c332980a07af1 without the 5th commit, because I don't think CBaseChainParams should move to the kernel. Comment about that below, but feel free to ignore the other comments.

Overall good changes, and nice that this is getting progressively simpler.

- [X] b56f0555fd2b048bb1d37474cda706efece49ed5 Decouple SigNetChainParams from ArgsManager (1/7)
- [X] 66b17894185e0cda3ea7cbec77cc771a7ba41d56 Decouple RegTestChainParams from ArgsManager (2/7)
- [X] 2ea0e3a4266f1b7c3bddec5d68cae9347a9aad57 Remove UpdateVersionBitsParameters (3/7)
- [X] e634311723aa43ba18bc45e1cf7e83fc4add5861 Add factory functions for Main/Test/Sig/Reg chainparams (4/7)
- [ ] 341be7fa680e9dcdbaa46db0194833b5d0229394 split non/kernel chainparamsbase (5/7)
- [X] 18099e9d28247002cc08407969feec4fada90b76 Split non/kernel chainparams (6/7)
- [X] c44983faa02b4845cd7605abe99c332980a07af1 refactor: Don't use global chainparams in chainstatemanager method (7/7)

TheCharlatan commented:
Thanks for the review,

Updated c44983faa02b4845cd7605abe99c332980a07af1 -> a7afa98ca38fd66fc69b77b95a6a57d50207911b ([tc/2022-09-libbitcoinkernel-chainparams-args_4](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_4) -> [tc/2022-09-libbitcoinkernel-chainparams-args_5](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_4..tc/2022-09-libbitcoinkernel-chainparams-args_5)) to address @ryanofsky 's latest comments. Drops the commit splitting chainparamsbase as suggested and addresses the two remaining nits.

TheCharlatan commented on src/chainparams.cpp:
```diff
@@ -497,41 +520,30 @@ class CRegTestParams : public CChainParams
-static void MaybeUpdateHeights(const ArgsManager& args, Consensus::Params& consensus)
-{
+void ReadRegTestArgs(const ArgsManager& args, CChainParams::RegTestOptions& options) {
+    options.fastprune = args.GetBoolArg("-fastprune", false);
```
Merging with the option seems more in tune with how the options are handled in other places in the code, so I went with that.

ryanofsky reviewed (approved):
Code review ACK a7afa98ca38fd66fc69b77b95a6a57d50207911b. This looks great, and much cleaner and simpler than the starting point. I'd encourage @ajtowns and anybody else who looked at this previously to have another look.

I think that one remaining drawback is that the PR description is very verbose. I think the PR could be described in a few sentences as:

* Move `CChainParams` class definition into the kernel, and give it new factory functions `CChainParams::{RegTest,Signet,Main,TestNet}` that allow it to be constructed without an `ArgsManager` reference, unlike the current factory function `CreateChainParams`.
  
  This allows external code using libbitcoinkernel to access to chain parameters and set chain options and call validation code without having to invoke bitcoind's argument parsing logic.
  
  Doing this required removing uses of `ArgsManager` within the `CChainParams` class itself, so the first few commits do that refactoring.

For reference, the [current PR description](https://github.com/bitcoin/bitcoin/pull/26177#issue-1384719233) is:

> This pull request is mostly a code-move of chainparams related functionality into the kernel library. @dongcarl is the original author of this patchset, these commits were taken from https://github.com/dongcarl/bitcoin/tree/2022-03-libbitcoinkernel-chainparams-args-only.
> 
> #### Context
> The bitcoin kernel library currently relies on code containing user configurations through the `ArgsManager`. This is not optimal, since as a stand-alone library it should not rely on most of the user-defined configuration. Instead, its interfaces should accept control and options structs that control the kernel library's desired configuration.
> 
> A lot of tweakable parameters are defined within `CChainParams` and depend on access on the global `ArgsManager gArgs` instance.
> 
> #### Changes
> This PR splits the chainparams code of chainparams.cpp/h and chainparamsbase.cpp/h into parts depending on `gArgs` access and parts that don't. These are then moved into separate files within the kernel/ directory.
> 
> Though a lot of code had to be moved and reshuffled to accommodate to the validation of both `gArgs` and options structs, the behavior should not change by the hand of this PR.
> 
> Similar work towards decoupling the `ArgsManager` from the kernel has been done in #25290, #25487 and #25527. Similar PR's are open in #25862 and #25152.
> 
> #### Currently Identified Drawbacks
> The patches touch a large amount of consensus critical code. I tried to keep the change sets as incremental and easy to review as possible, but a lot of manual comparison of diffs will have to be done.
> 
> There is no patch done yet introducing the kernel/chainparams.cpp/h into the kernel namespace. Such a change would be very intrusive to the entire codebase and hard to review as part of this patch set. Due to its intrusiveness, I am not sure whether this is the correct path forward or not.
> 
> ~The `globalChainBaseParams` global had to be moved into the kernel chainparamsbase.cpp in order to allow calls to `BaseParams()` from contexts without access to gArgs. This is not ideal, but was the only way forward I could think of so far without basing this pull request on more prior work decoupling `gArgs` from the various call sites of `BaseParams()`.~ ~EDIT: `chainparamsbase.cpp` is left untouched now. Only the class declaration of `CBaseChainParams` is moved to the kernel directory.~ EDIT 2: `chainparamsbase.*` is left untouched. See [#26177 (comment)](https://github.com/bitcoin/bitcoin/pull/26177#discussion_r1130039156) for reasoning.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/26177#issue-1384719233

> There is no patch done yet introducing the kernel/chainparams.cpp/h into the kernel namespace. Such a change would be very intrusive to the entire codebase and hard to review as part of this patch set. Due to its intrusiveness, I am not sure whether this is the correct path forward or not.

I think this PR is doing the right thing. If you move code into the kernel incrementally, and then set the namespace at the end after the code is moved, that avoids the need in incremental commits to add `kernel::` in code before it is moved, and then remove `kernel::` after moving it, which would just be a lot of noise.

TheCharlatan commented:
re: https://github.com/bitcoin/bitcoin/pull/26177#pullrequestreview-1332732743

> I think that one remaining drawback is that the PR description is very verbose. I think the PR could be described in a few sentences as:
...

Thank you for the suggestions. I cleaned the description.

TheCharlatan commented:
Rebased a7afa98ca38fd66fc69b77b95a6a57d50207911b -> 45048c0d7e0ee12210d226239fb5ce63d9ed4441 ([tc/2022-09-libbitcoinkernel-chainparams-args_5](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_5) -> [tc/2022-09-libbitcoinkernel-chainparams-args_6](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_6), [compare](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_5..tc/2022-09-libbitcoinkernel-chainparams-args_6)) to resolve conflict.

ajtowns commented on src/bitcoin-chainstate.cpp:
```diff
@@ -52,7 +53,7 @@ int main(int argc, char* argv[])
     // SETUP: Misc Globals
     SelectParams(CBaseChainParams::MAIN);
-    const CChainParams& chainparams = Params();
+    auto chainparams = CChainParams::Main();
```
I don't follow why this is being changed when `SelectParams()` is still used -- `SelectParams` already relies on `gArgs`  and creates its own `unique_ptr<CChainParams>` so this is just creating two copies of the same object?

TheCharlatan replied:
This is a demo binary, so I think the point is to show that the `chainman` can now be instantiated with a `chainparams` that is itself instantiated without directly relying on `gArgs`. I'm not sure if this really makes sense though.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/26177#discussion_r1135671989

> This is a demo binary, so I think the point is to show that the `chainman` can now be instantiated with a `chainparams` that is itself instantiated without directly relying on `gArgs`. I'm not sure if this really makes sense though.

IMO, it would be a little clearer if the vague "SETUP: Misc Globals" comment was changed to say specifically that SelectParams() call was needed to set some globals used by validation code and could be removed when the globals were no longer needed.

ajtowns commented on src/kernel/chainparams.cpp:
```diff
@@ -0,0 +1,531 @@
+ */
+class SigNetParams : public CChainParams {
+public:
+    explicit SigNetParams(const CChainParams::SigNetOptions& options)
```
Closer to a move-only commit if you didn't add the `CChainParams::` qualifier here (and for regtest)

ajtowns reviewed (commented):
ACK 45048c0d7e0ee12210d226239fb5ce63d9ed4441

Agree with ryanofsky -- "This looks great, and much cleaner and simpler than the starting point." but some nits anyway.

ryanofsky reviewed (approved):
Code review ACK 600ab02bf58e073a93936438a7e884b3a7110f1c. Only changes since last review were rebase, and making suggested changes to make the split commit move only

ajtowns reviewed (approved):
ACK 600ab02bf58e073a93936438a7e884b3a7110f1c

maflcko reviewed (approved):
Nice. I left some non-blocking nits. Please advise maintainers if this should be merged and if you plan to address the nits (if any).

review ACK 600ab02bf58e073a93936438a7e884b3a7110f1 💒

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 600ab02bf58e073a93936438a7e884b3a7110f1 💒
kzm7rHrxddPB+LjaomzYoCJON/6jVUm4ilbYIF+uPjDG2kXx6jB3bd9r+RtMuVEKCtA3jrp/uE0wa/yO1RytDA==
```

</details>

TheCharlatan commented:
re https://github.com/bitcoin/bitcoin/pull/26177#pullrequestreview-1341271390
>  Please advise maintainers if this should be merged and if you plan to address the nits (if any).

Yes, will address the nits. It think it's worth another round.

TheCharlatan commented on src/chainparams.cpp:
```diff
@@ -382,11 +378,25 @@ class SigNetParams : public CChainParams {
+    if (args.IsArgSet("-signetchallenge")) {
+        const auto signet_challenge = args.GetArgs("-signetchallenge");
+        if (signet_challenge.size() != 1) {
+            throw std::runtime_error(strprintf("%s: -signetchallenge cannot be multiple values.", __func__));
```
This produces a log like:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  CreateChainParams: Unknown chain lol
```
So I think yes, it does add value.

TheCharlatan commented:
Thank you for the review!

Updated 600ab02bf58e073a93936438a7e884b3a7110f1c -> b3e78dc91d01e364b77aacd9fb9a2f88688ab8a6 ([tc/2022-09-libbitcoinkernel-chainparams-args_7](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_7) -> [tc/2022-09-libbitcoinkernel-chainparams-args_8](https://github.com/TheCharlatan/bitcoin/commits/tc/2022-09-libbitcoinkernel-chainparams-args_8), [compare](https://github.com/TheCharlatan/bitcoin/compare/tc/2022-09-libbitcoinkernel-chainparams-args_7..tc/2022-09-libbitcoinkernel-chainparams-args_8)) to address @MarcoFalke's nits.

ryanofsky reviewed (approved):
Code review ACK b3e78dc91d01e364b77aacd9fb9a2f88688ab8a6. Only changes since last review were recent review suggestions.

ajtowns reviewed (commented):
ACK b3e78dc91d01e364b77aacd9fb9a2f88688ab8a6
